rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java

```diff
@@ -0,0 +1,262 @@
+package co.rsk.peg.bitcoin;
+
+import co.rsk.bitcoinj.core.Address;
+import co.rsk.bitcoinj.core.BtcECKey;
+import co.rsk.bitcoinj.core.BtcTransaction;
+import co.rsk.bitcoinj.core.Coin;
+import co.rsk.bitcoinj.core.NetworkParameters;
+import co.rsk.bitcoinj.core.Sha256Hash;
+import co.rsk.bitcoinj.script.FastBridgeP2shErpRedeemScriptParser;
+import co.rsk.bitcoinj.script.Script;
+import co.rsk.bitcoinj.script.ScriptBuilder;
+import co.rsk.bitcoinj.wallet.RedeemData;
+import co.rsk.config.BridgeConstants;
+import co.rsk.config.BridgeMainNetConstants;
+import co.rsk.peg.Federation;
+import co.rsk.peg.FederationTestUtils;
+import co.rsk.peg.P2shErpFederation;
+import co.rsk.peg.PegTestUtils;
+import org.ethereum.config.blockchain.upgrades.ActivationConfig;
+import org.ethereum.config.blockchain.upgrades.ActivationConfigsForTest;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+class BitcoinUtilsTest {
+    private static final BridgeConstants bridgeMainnetConstants = BridgeMainNetConstants.getInstance();
+    private static final NetworkParameters btcMainnetParams = bridgeMainnetConstants.getBtcParams();
+    private static final int FIRST_OUTPUT_INDEX = 0;
+    private static final int FIRST_INPUT_INDEX = 0;
+
+    List<BtcECKey> signers;
+    List<BtcECKey> pubKeys;
+    Address destinationAddress;
+
+    @BeforeEach
+    void init() {
+        signers = BitcoinTestUtils.getBtcEcKeysFromSeeds(
+            new String[]{"fa01", "fa02", "fa03"}, true
+        );
+        pubKeys = signers.stream().map(BtcECKey::getPubKeyPoint).map(BtcECKey::fromPublicOnly)
+            .collect(Collectors.toList());
+
+        destinationAddress = BitcoinTestUtils.createP2PKHAddress(btcMainnetParams, "destinationAddress");
+    }
+
+    @Test
+    void test_getFirstInputSigHash_from_multisig() {
+        // Arrange
+        int totalSigners = signers.size();
+        int requiredSignatures = totalSigners / 2 + 1;
+
+        Script redeemScript = ScriptBuilder.createRedeemScript(requiredSignatures, signers);
+
+        Address destinationAddress = BitcoinTestUtils.createP2PKHAddress(btcMainnetParams, "destinationAddress");
+
+        BtcTransaction btcTx = new BtcTransaction(btcMainnetParams);
+        btcTx.addInput(BitcoinTestUtils.createHash(1), FIRST_OUTPUT_INDEX, new Script(new byte[]{}));
+        btcTx.addOutput(Coin.COIN, destinationAddress);
+
+        RedeemData redeemData = RedeemData.of(pubKeys, redeemScript);
+        Script p2SHScript = ScriptBuilder.createP2SHOutputScript(requiredSignatures, pubKeys);
+        Script inputScript = p2SHScript.createEmptyInputScript(redeemData.keys.get(0), redeemData.redeemScript);
```

Comment:
```suggestion
        Script p2SHScript = ScriptBuilder.createP2SHOutputScript(redeemScript);
        Script inputScript = p2SHScript.createEmptyInputScript(null, redeemData.redeemScript);
        btcTx.getInput(FIRST_INPUT_INDEX).setScriptSig(inputScript);
```
See the implementation for createEmptyInputScript, you either need the key if you are creating a script sig that pays to a key, or the redeem script if you are creating a script sig that pays to a script. Not both.

And I guess this will be quite commonly used so it could be extracted to it's own method. Is it only used in tests? Or would it make sense to have it in BitcoinUtils class?

Reply:
I'd check usages in the PowPeg node also